File: apps/web/src/types/api/reader-notes.ts
import type { WebAnchorSegmentVm } from "./annotations";
import type { UserEditorialAssetAnchorDto } from "./reader-plate";

export interface ReaderNoteCreateRequestDto {
  // DATA-LEGACY-IDENTITY-EXIT: the Reading Record anchor is the only note
  // contract accepted by the API.
  anchor: UserEditorialAssetAnchorDto;
  selected_text: string;
  note_text: string;
  payload_json?: Record<string, unknown>;
}

export interface ReaderNoteUpdateRequestDto {
  note_text: string;
}

export interface ReaderNoteResponseDto {
  id: string;
  quote_mode: "sentence" | "text_range" | "multi_text";
  target_key: string;
  paragraph_id: string | null;
  sentence_id: string | null;
  selected_text: string;
  start_offset: number | null;
  end_offset: number | null;
  text_hash: string | null;
  segments: Array<{
    paragraph_id?: string | null;
    sentence_id: string;
    selected_text: string;
    start_offset: number;
    end_offset: number;
    text_hash: string;
  }>;
  note_text: string;
  payload_json: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  reading_record_id?: string | null;
  base_id?: string | null;
  generation?: number | null;
  unit_id?: string | null;
  anchor_segment_id?: string | null;
  unit_start_utf16?: number | null;
  unit_end_utf16?: number | null;
}

export interface ReaderNoteListResponseDto {
  items: ReaderNoteResponseDto[];
}

export interface WebReaderNoteVm {
  id: string;
  recordId: string;
  anchorSentenceId: string;
  quoteMode: "sentence" | "text_range" | "multi_text";
  targetKey: string;
  paragraphId: string | null;
  sentenceId: string | null;
  selectedText: string;
  startOffset: number | null;
  endOffset: number | null;
  textHash: string | null;
  segments: WebAnchorSegmentVm[];
  noteText: string;
  createdAt: string;
  updatedAt: string;
}

export interface WebReaderNoteCreateRequest {
  recordId: string;
  quoteMode: "sentence" | "text_range" | "multi_text";
  anchorSentenceId: string;
  paragraphId?: string;
  sentenceId?: string;
  selectedText: string;
  startOffset?: number | null;
  endOffset?: number | null;
  textHash?: string | null;
  segments?: WebAnchorSegmentVm[];
  noteText: string;
  payloadJson?: Record<string, unknown>;
}

File: services/api/app/schemas/reader_notes.py
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from app.schemas.user_annotations import UserAnnotationSegment
from app.schemas.user_editorial_assets import UserEditorialAssetAnchor


class ReaderNoteCreateRequest(BaseModel):
    # DATA-LEGACY-IDENTITY-EXIT: the Reading Record anchor is the only note
    # contract; the Reading Record anchor gate is the sole validation
    # authority. Legacy analysis_record / render_scene fields are gone.
    anchor: UserEditorialAssetAnchor
    selected_text: str
    note_text: str = Field(min_length=1)
    payload_json: dict = Field(default_factory=dict)

    @model_validator(mode="after")
    def validate_quote_fields(self):
        if not self.selected_text.strip():
            raise ValueError("selected_text must not be empty")
        if self.selected_text != self.anchor.selected_text:
            raise ValueError("selected_text must match anchor.selected_text")
        if not self.note_text.strip():
            raise ValueError("note_text must not be empty")
        return self


class ReaderNoteUpdateRequest(BaseModel):
    note_text: str = Field(min_length=1)


class ReaderNoteResponse(BaseModel):
    id: UUID
    quote_mode: str
    target_key: str
    paragraph_id: str | None = None
    sentence_id: str | None = None
    selected_text: str
    start_offset: int | None = None
    end_offset: int | None = None
    text_hash: str | None = None
    segments: list[UserAnnotationSegment] = Field(default_factory=list)
    note_text: str
    payload_json: dict
    created_at: str
    updated_at: str
    # D6-U4 Reading Record anchor columns — the only anchor identity.
    reading_record_id: UUID | None = None
    base_id: UUID | None = None
    generation: int | None = None
    unit_id: str | None = None
    anchor_segment_id: str | None = None
    unit_start_utf16: int | None = None
    unit_end_utf16: int | None = None


class ReaderNoteListResponse(BaseModel):
    items: list[ReaderNoteResponse]
